cmd/remediation-aggregator/main.go

```diff
@@ -0,0 +1,426 @@
+// for now (?) just diffs the MC specs and the remediation type, not sure if we'll ever want to diff more
+func diffRemediations(old, new *complianceoperatorv1alpha1.ComplianceRemediation) bool {
+	if old.Spec.Type != new.Spec.Type {
+		return false
+	}
+
+	// should we be more picky and just compare what can be set with the remediations? e.g. OSImageURL can't
+	// be set with a remediation..
+	return reflect.DeepEqual(old.Spec.MachineConfigContents.Spec, new.Spec.MachineConfigContents.Spec)
+}
+
+func updateResultsConfigMap(clientset *kubernetes.Clientset, cm *v1.ConfigMap) error {
+	cmCopy := cm.DeepCopy()
+
+	if cmCopy.Annotations == nil {
+		cmCopy.Annotations = make(map[string]string)
+	}
+	cmCopy.Annotations[configMapRemediationsProcessed] = ""
+
+	_, err := clientset.CoreV1().ConfigMaps(cmCopy.Namespace).Update(cmCopy)
+	return err
+}
+
+func createRemediations(crClient *complianceCrClient, scan *complianceoperatorv1alpha1.ComplianceScan, remediations []*complianceoperatorv1alpha1.ComplianceRemediation) error {
+	for _, rem := range remediations {
+		fmt.Printf("Creating remediation %s\n", rem.Name)
+		if err := controllerutil.SetControllerReference(scan, rem, crClient.scheme); err != nil {
+			fmt.Printf("Failed to set remediation ownership %v", err)
+			return err
+		}
+
+		if rem.Labels == nil {
+			rem.Labels = make(map[string]string)
+		}
+		rem.Labels[complianceoperatorv1alpha1.ScanLabel] = scan.Name
+		rem.Labels[complianceoperatorv1alpha1.SuiteLabel] = scan.Labels["compliancesuite"]
+		rem.Labels[mcfgv1.McRoleKey] = getScanRoleLabel(scan.Spec.NodeSelector)
+		if rem.Labels[mcfgv1.McRoleKey] == "" {
+			return fmt.Errorf("scan %s has no role assignment", scan.Name)
+		}
+
+		err := crClient.client.Create(context.TODO(), rem)
+		if err != nil {
+			fmt.Printf("Failed to create remediation: %v\n", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func readContent(filename string) (string, error) {
+	// gosec complains that the file is passed through an evironment variable. But
+	// this is not a security issue because none of the files are user-provided
+	cleanFileName := filepath.Clean(filename)
+	// #nosec G304
+	file, err := os.Open(cleanFileName)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	b, err := ioutil.ReadAll(file)
+	return string(b), err
+}
+
+func aggregator(cmd *cobra.Command, args []string) {
+	var scanRemediations []*complianceoperatorv1alpha1.ComplianceRemediation
+
+	aggregatorConf := parseConfig(cmd)
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		// fallback to kubeconfig
+		fmt.Println("Running outside cluster")
+		kubeconfig := filepath.Join("~", ".kube", "config")
+		if envvar := os.Getenv("KUBECONFIG"); len(envvar) > 0 {
+			kubeconfig = envvar
+		}
+		fmt.Printf("Using KUBECONFIG %s\n", kubeconfig)
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			fmt.Printf("The kubeconfig cannot be loaded: %v\n", err)
+			os.Exit(1)
+		}
+	}
```

Comment:
Will this ever run outside the cluster though? I think you only need the InClusterconfig call, and fail if that doesn't work.

Reply:
I added this code for easier development initially so I can just create some objects without running all the scans and then run the binary locally. I removed the code now.